src/fastcs/backends/rest/rest.py

```diff
@@ -0,0 +1,158 @@
+    )
+
+
+def _wrap_attr_get(
+    attribute: AttrR[T],
+) -> Callable[[], Coroutine[Any, Any, Any]]:
+    async def attr_get() -> Any:  # Must be any as response_model is set
+        value = attribute.get()  # type: ignore
+        return {"value": value}
+
+    return attr_get
+
+
+def _add_attribute_api_routes(app: FastAPI, mapping: Mapping) -> None:
+    for single_mapping in mapping.get_controller_mappings():
+        path = single_mapping.controller.path
+
+        for attr_name, attribute in single_mapping.attributes.items():
+            attr_name = attr_name.replace("_", "-")
+            route = f"{'/'.join(path)}/{attr_name}" if path else attr_name
+
+            match attribute:
+                # https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
+                case AttrRW():
+                    app.add_api_route(
+                        f"/{route}",
+                        _wrap_attr_get(attribute),
+                        methods=["GET"],  # Idempotent and safe data retrieval,
+                        status_code=200,  # https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET
+                        response_model=_get_response_body(attribute),
+                    )
+                    app.add_api_route(
+                        f"/{route}",
+                        _wrap_attr_put(attribute),
+                        methods=["PUT"],  # Idempotent state change
+                        status_code=204,  # https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT
+                    )
+                case AttrR():
+                    app.add_api_route(
+                        f"/{route}",
+                        _wrap_attr_get(attribute),
+                        methods=["GET"],
+                        status_code=200,
+                        response_model=_get_response_body(attribute),
+                    )
+                case AttrW():
+                    app.add_api_route(
+                        f"/{route}",
+                        _wrap_attr_put(attribute),
+                        methods=["PUT"],
+                        status_code=204,
+                    )
+
+
+def _wrap_command(
+    method: Callable, controller: BaseController
+) -> Callable[..., Awaitable[None]]:
+    async def command() -> None:
+        await getattr(controller, method.__name__)()
+
+    return command
+
+
+def _add_command_api_routes(app: FastAPI, mapping: Mapping) -> None:
+    for single_mapping in mapping.get_controller_mappings():
+        path = single_mapping.controller.path
+
+        for name, method in single_mapping.command_methods.items():
+            cmd_name = name.replace("_", "-")
+            route = f"/{'/'.join(path)}/{cmd_name}" if path else cmd_name
+            app.add_api_route(
+                f"/{route}",
+                _wrap_command(
+                    method.fn,
+                    single_mapping.controller,
+                ),
+                methods=["PUT"],
```

Comment:
Should commands use POST?

Reply:
Ah

> A PUT request is used over a POST as the action is idempotent

Could you expand on this?

Reply:
Good to discuss

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST
```
The difference between [PUT] and POST is that PUT is idempotent, 
calling it once is no different from calling it several times successively (there are no side effects). 
Successive identical POST requests may have additional effects, 
such as creating the same order several times.
```

Here a more concrete example:

https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/PUT-vs-POST-Whats-the-difference
```
For example, I perform a PUT operation to set my bank account balance to $100. 
Even if I perform that operation 100 times, the end result is a balance of $100.

In contrast, if I perform 100 POST operations that each increases my account balance by 10%, 
the result of each interaction creates a different result. That is not an idempotent operation.

A PUT method allows only idempotent operations. 
POST methods are not subject to idempotency requirements."
```
By this definition I consider a command is a PUT

A POST could be "safe" in a sense that a command may not be idempotent. A big distinction seems to be that a POST must return a body - which is not something are doing in EPICS...

Reply:
Good timing! Fair enough. Of course I will be biased but I like having `/subcontroller/{method or attribute}` as the route so that we can see what is available from the docs and having very simple schemas and logic. Do you have other suggestions?

Reply:
Is there a precedent in FastAPI that we can copy here?

Reply:
FastAPI does not constrain you to using a RestAPI, so we should just try be restful which is why I refer to the mozilla docs. An interesting reference is possibly "https://tango-controls.readthedocs.io/en/latest/development/advanced/rest-api.html" they do use an empty put for commands, but they support 1 arg

Reply:
Note the API we use is fully described at `/docs` by swagger

Reply:
OK. With odin I think we are just going with PUT as well, so that seems reasonable.